packages/contracts/contracts/libs/v0.8.x/minter-libs/MaxInvocationsLib.sol

```diff
@@ -234,4 +237,126 @@ library MaxInvocationsLib {
+    ) internal view returns (bool) {
+        // get max invocations from core contract
+        uint256 coreInvocations;
+        uint256 coreMaxInvocations;
+        (coreInvocations, coreMaxInvocations) = MaxInvocationsLib
+            .coreContractInvocationData(_projectId, _coreContract);
+        uint256 localMaxInvocations = _maxInvocationsProjectConfig
+            .maxInvocations;
+        // value is locally defined, and could be out of date.
+        // only possible illogical state is if local max invocations is
+        // greater than core contract's max invocations, in which case
+        // we should use the core contract's max invocations
+        if (localMaxInvocations > coreMaxInvocations) {
+            // local max invocations is stale and illogical, defer to core
+            // contract's max invocations since it is the limiting factor
+            return (coreMaxInvocations == coreInvocations);
+        }
+        // local max invocations is limiting, so check core invocations against
+        // local max invocations
+        return (coreInvocations >= localMaxInvocations);
+    }
+
+    /**
+     * @notice Verifies the cached values of a project's maxInvocation state
+     * are logically consistent with the core contract's maxInvocation state,
+     * or populates them to equal the core contract's maxInvocation state if
+     * they have never been populated or are illogical (e.g. minter max
+     * invocations are greater than core contract max invocations).
+     * @dev intended for use with settlement lib
+     * @dev assumes core contract's max invocations may only be reduced, which
+     * is the case for V3 core contracts
+     */
+    function refreshMaxInvocations(
+        uint256 _projectId,
+        address _coreContract,
+        MaxInvocationsProjectConfig storage _maxInvocationsProjectConfig
+    ) internal {
+        // project's max invocations and has max been invoked can only be
+        // initial values if never populated, because setting a maxInvocations
+        // of zero means maxHasBeenInvoked would be set to true
+        bool notPopulated = (_maxInvocationsProjectConfig.maxInvocations == 0 &&
+            _maxInvocationsProjectConfig.maxHasBeenInvoked == false);
+        if (notPopulated) {
+            // sync the minter max invocation state to equal the values on the
+            // core contract (least restrictive state)
+            MaxInvocationsLib.syncProjectMaxInvocationsToCore(
+                _projectId,
+                _coreContract,
+                _maxInvocationsProjectConfig
+            );
+        } else {
+            // if using local max invocations, validate the local state
+            // (i.e. ensure local max invocations not greater than core max
+            // invocations)
+            _validateProjectMaxInvocations(
+                _projectId,
+                _coreContract,
+                _maxInvocationsProjectConfig
+            );
+        }
```

Comment:
I think I understand what you are saying here, but just wanted to note that the `_validateProjectMaxInvocations` also handles the case where local max invocations > core max invocations (e.g. artist reduced max invocations on core)

Reply:
This was refactored a bit, and comments hopefully made much clearer in [3bebf8e](https://github.com/ArtBlocks/artblocks-contracts/pull/928/commits/3bebf8e730a1b56efd90c5e50f0b29d2ff77697c).

Additionally, I caught that events were not properly being emitted after potential state change of max invocations in three places where the refresh function was called, so that was fixed in the same comment (+tests added)

Reply:
Cc @lyaunzbe @jakerockland on the previous comment ^^

Reply:
refactor looks good + ty for catching those missed events